### PR TITLE
Adding tooling utils and cleaning up VM context/module creation.

### PR DESCRIPTION
**Pull request by benvanik**
This starts to centralize tooling logic to allow for more flexible module composition. Combined with a new `--iree-execution-model=` compiler flag it's now possible to not use the HAL - and in subsequent PRs use variants of the HAL/VMVX/etc directly from the host module.

Eventually moving module dependency resolution into the core APIs (`iree_vm_context_resolve_modules`) will make it possible to dynamically load multiple transitively defined modules based on the main module provided. For now this keeps things simple.

Part of this work required reordering type registration. Lots of places were playing fast and loose with ordering of type registration and module creation and now these require an `instance` arg so things must be ordered (even if the type registration is not yet done per-instance as in #8698).

The python `VmModule.from_flatbuffer` signature has been changed to take `(vm_instance, file)`.

### CODE REVIEW

File: samples/py_custom_module/decode_secret_message.py [hunk 1]
```diff
@@ -96,17 +96,17 @@ def jointokens(self, token_list_ref):
...
-  return rt.VmModule.from_flatbuffer(vmfb_contents)
+  return compiler.tools.compile_file(os.path.join(os.path.dirname(__file__),
+                                                  "main.mlir"),
+                                     target_backends=["vmvx"])
 
 
 def main():
   print("Compiling...")
-  main_module = compile()
+  vmfb_contents = compile()
```
Comment (stellaraccident) [inline samples/py_custom_module/decode_secret_message.py:106]: Note: I need to flip this around and rework the JIT APIs so this is fine for now.

**Review by stellaraccident — CHANGES_REQUESTED** [review]
Marking request changes because there is a lifetime question. Looks good otherwise, and glad we are fixing this (it will only get harder).

File: runtime/bindings/python/vm.cc [hunk 2]
```diff
@@ -58,9 +58,17 @@ py::dict GetFunctionReflectionDict(iree_vm_function_t& f) {
...
-  iree_vm_instance_t* instance;
+
+  iree_vm_instance_t* instance = NULL;
   auto status = iree_vm_instance_create(iree_allocator_system(), &instance);
   CheckApiStatus(status, "Error creating instance");
+
+  // The python bindings assume the HAL is always available for use.
+  // We register the types here so modules can be loaded using the HAL types
+  // in any order.
+  CheckApiStatus(iree_hal_module_register_all_types(instance),
```
[the remark this answers is not in the document]
Comment (benvanik) [reply runtime/bindings/python/vm.cc:69]: The types are required to load modules (though with some work we could probably change that) and are associated with an instance that declares them and that's mostly based on whether the types are compiled into the binary. The python bindings pull in the HAL module and related code unconditionally so there's no harm in registering all available (you can't do anything with the types on their own). If we wanted to have a python build configuration without the full HAL (or any of the HALs at all) compiled in we could make it a compile flag but I'm not sure we care too much.

I'm going to make sure that bytecode modules (currently the place we memoize types during module loading) properly error if types aren't registered. That'll make it easier for users to see if their own custom types aren't registered before they try to load a module using them.

File: runtime/bindings/python/hal.cc [hunk 3]
```diff
@@ -403,8 +403,6 @@ VmModule CreateHalModule(HalDevice* device) {
 void SetupHalBindings(pybind11::module m) {
   py::dict driver_cache;
 
-  IREE_CHECK_OK(iree_hal_module_register_all_types());
-
   // Built-in module creation.
   m.def("create_hal_module", &CreateHalModule);
```
[the remark this answers is not in the document]
Comment (benvanik) [reply runtime/bindings/python/hal.cc:407]: The HAL module does retain the device and if the modules need the instance they'll retain it (today they don't and it's fine if it disappears), so I think we're good? I hadn't considered the retaining and that may explain some of the CI failures. I'll take a better look and see if I can spot anything bad.

**Comment by benvanik** [comment]
Any idea what this means?
https://github.com/iree-org/iree/runs/7707716817?check_suite_focus=true

```
File "/home/runner/work/iree/iree/runtime/bindings/python/iree/runtime/system_api.py", line 54, in Config: No attribute 'HalDevice' on module 'runtime._binding' [module-attr]
File "/home/runner/work/iree/iree/runtime/bindings/python/iree/runtime/system_api.py", line 55, in Config: No attribute 'VmInstance' on module 'runtime._binding' [module-attr]
File "/home/runner/work/iree/iree/runtime/bindings/python/iree/runtime/system_api.py", line 56, in Config: No attribute 'VmModule' on module 'runtime._binding' [module-attr]
...
```

I don't think I changed anything that would cause that?

File: runtime/bindings/python/vm.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (benvanik) [reply runtime/bindings/python/vm.cc:69]: Ok, made it a bytecode module load failure to not have types registered - should make it easier to catch issues.

File: runtime/bindings/python/hal.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (benvanik) [reply runtime/bindings/python/hal.cc:407]: CI seems fine (issues were unrelated).

File: docs/developers/developing_iree/developer_overview.md [hunk 4]
```diff
@@ -180,10 +180,11 @@ For example, to execute the contents of
 [samples/models/simple_abs.mlir](https://github.com/iree-org/iree/blob/main/samples/models/simple_abs.mlir):
 
 ```shell
+# iree-run-mlir <compiler flags> [input.mlir] <runtime flags>
 $ ../iree-build/tools/iree-run-mlir \
-  --function-input="f32=-2" \
   --iree-hal-target-backends=vmvx \
-  $PWD/samples/models/simple_abs.mlir
+  $PWD/samples/models/simple_abs.mlir \
+  --function_input=f32=-2
```
Comment (ScottTodd) [inline docs/developers/developing_iree/developer_overview.md:187]: Thank you for this cleanup!

**Review by ScottTodd — APPROVED** [review]

File: docs/developers/developing_iree/developer_overview.md (hunk 4 above)
Comment (benvanik) [reply docs/developers/developing_iree/developer_overview.md:187]: sorry it took so long! had been procrastinating on that for years :)

File: runtime/bindings/python/hal.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (stellaraccident) [reply runtime/bindings/python/hal.cc:407]: Yes - then we're good!

**Review by stellaraccident — APPROVED** [review]

**Pull request merged**